File: python/examples/all_clustering.py
#!/usr/bin/env python
"""
Explicit examples on how to use clustering
"""

from numpy import array, concatenate, sign
from numpy.random import rand, seed, permutation
from sg import sg

def get_clouds (num, num_feats, num_vec):
	data=[rand(num_feats, num_vec)+x/2 for x in xrange(num)]
	cloud=concatenate(data, axis=1)
	return array([permutation(x) for x in cloud])

def kmeans ():
	print 'KMeans'

	size_cache=10
	k=3
	iter=1000
	num_feats=14
	num_trainvec=10

	trainlab=sign(rand(1, num_trainvec*k)-0.5)[0]
	traindata=get_clouds(k, num_feats, num_trainvec)

	sg('set_features', 'TRAIN', traindata)
	sg('set_labels', 'TRAIN', trainlab)
	sg('set_distance', 'EUCLIDIAN', 'REAL')
	sg('init_distance', 'TRAIN')
	sg('new_classifier', 'KMEANS')
	sg('train_classifier', k, iter)

	[radi, centers]=sg('get_classifier')


def hierarchical ():
	print 'Hierarchical'

	size_cache=10
	merges=3
	num_feats=14
	num_trainvec=10

	traindata=get_clouds(merges, num_feats, num_trainvec)

	sg('set_features', 'TRAIN', traindata)
	sg('set_distance', 'EUCLIDIAN', 'REAL')
	sg('init_distance', 'TRAIN')
	sg('new_classifier', 'HIERARCHICAL')
	sg('train_classifier', merges)

	[merge_distance, pairs]=sg('get_classifier')

###########################################################################
# call functions
###########################################################################

if __name__=='__main__':
	kmeans()
	hierarchical()

File: r/examples/all_clustering.R
# Explicit examples on how to use clustering
#
# run as R --no-save --slave --file=<filename>

library("sg")
#uncomment if make install does not work and comment the library("sg") line above
#dyn.load('sg.so')
#sg <- function(...) .External("sg",...,PACKAGE="sg")

len <- 12
num <- 30
size_cache <- 10

traindat <- matrix(c(rnorm(len*num)-1,rnorm(len*num)+1), len, 2*num)
trainlab <- c(rep(-1,num),rep(1,num))


# KMEANS
print('KMeans')

k <- 3
iter <- 1000

dump <- sg('set_distance', 'EUCLIDIAN', 'REAL')
dump <- sg('set_features', 'TRAIN', traindat)
dump <- sg('set_labels', 'TRAIN', trainlab)
dump <- sg('init_distance', 'TRAIN')
dump <- sg('new_classifier', 'KMEANS')
dump <- sg('train_classifier', k, iter)

result <- sg('get_classifier')
radi <- result[[1]]
centers <- result[[2]]


# Hierarchical
print('Hierarchical')

merges=3

dump <- sg('set_features', 'TRAIN', traindat)
dump <- sg('set_distance', 'EUCLIDIAN', 'REAL')
dump <- sg('init_distance', 'TRAIN')
dump <- sg('new_classifier', 'HIERARCHICAL')
dump <- sg('train_classifier', merges)

result <- sg('get_classifier')
merge_distances <- result[[1]]
pairs <- result[[2]]
